# 

to pull a specified region from a fasta file with one line 

```
# chrII:363400-364449

awk 'BEGIN{RS=">";FS="\n"}NR>1{if ($1~/chrII/)print ">"$0}'  genome.fasta | awk 'BEGIN{RS =">"; FS="\n"}NR>1{seq="";for (i=2;i<=NF;i++) seq=seq""$i; print ">"$1"\n"substr(seq,363400,364449-363400+1)}'
```

Following code is to change the chromosome names of the vcf files found in the path that is generated by the software snp-sites from integer to the the style "chr\[I-X]".

```
for i in EC1118_chr*_EC1118_*.vcf;do chr=`echo $i |sed 's/_/  /g'|cut -f2`; outfile=`echo $i | sed 's/.vcf/_chr/g'`;echo $chr $outfile;head $i -n4 > $outfile;tail $i -n+5 | awk -v chr="$chr" '$0=gensub(/1/, chr, $1)' >> $outfile; done
```

Count the each scaffold in a fasta file
```
awk '/^>/ {if (seqlen){print seqlen}; printf $0"\t";seqlen=0;next; } { seqlen += length($0)}END{print seqlen}' sequences.fa
```

Split the chromosomes
```
awk -v r=Anchorman_Ale  -F "|" '/^>/ {close(F); ID=$1; gsub("^>", "", ID); F=r"_"ID".fa"; gsub(" ", "", F)} {print >> F}' Anchorman_Ale_purged.fa
```
